tests/asgi/test_buffered_reader.py

```diff
@@ -0,0 +1,387 @@
+import io
+
+import pytest
+
+from falcon import testing
+from falcon.asgi import reader
+from falcon.errors import OperationNotAllowed
+
+
+async def async_iter(data):
+    for item in data:
+        yield item
+
+
+def async_take(source, count=None):
+    async def collect():
+        result = []
+        async for item in source:
+            result.append(item)
+            if count is not None and count <= len(result):
+                return result
+        return result
+
+    return testing.invoke_coroutine_sync(collect)
+
+
+async def chop_data(data, min_size=1024, max_size=64 * 1024):
+    index = 0
+    size = min_size
+
+    while True:
+        chunk = data[index:index + size]
+        index += size
+        if not chunk:
+            break
+        yield chunk
+
+        size = min(max_size, size + 1)
+
+
+SOURCE1 = (
+    b'Hello',
+    b'',
+    b'',
+    b',',
+    b' ',
+    b'World!',
+    b'\n',
+    b'Jus',
+    b't tes',
+    b'ting some iterato',
+    b'r goodne',
+    b'',
+    b'ss.',
+    b'',
+    b'\n',
```

Comment:
Aren't we missing non printable / non achi chars tests here?

Reply:
We are.
But, OTOH, multipart tests include these, including parsing completely random binary forms. Both in "unit" tests, and against a real `uvicorn` server.